stackdriver/config/config_test.go

```diff
@@ -0,0 +1,97 @@
+package config
+
+import (
+	corev1 "k8s.io/api/core/v1"
+
+	"testing"
+)
+
+func TestNewStackdriverConfigFromConfigMap(t *testing.T) {
+	tests := []struct {
+		name           string
+		configMap      *corev1.ConfigMap
+		expectedConfig Config
+	}{
+		{
+			name: "fullSdConfig",
+			configMap: &corev1.ConfigMap{
+				Data: map[string]string{
+					"project-id":           "project",
+					"gcp-location":         "us-west1",
+					"cluster-name":         "cluster",
+					"gcp-secret-name":      "secret",
+					"gcp-secret-namespace": "non-default",
+				},
+			},
+			expectedConfig: Config{
+				ProjectID:          "project",
+				GCPLocation:        "us-west1",
+				ClusterName:        "cluster",
+				GCPSecretName:      "secret",
+				GCPSecretNamespace: "non-default",
+			},
+		},
+		{
+			name:           "emptySdConfig",
+			configMap:      &corev1.ConfigMap{},
+			expectedConfig: Config{},
+		},
+		{
+			name: "partialSdConfig",
+			configMap: &corev1.ConfigMap{
+				Data: map[string]string{
+					"project-id":   "project",
+					"gcp-location": "us-west1",
+					"cluster-name": "cluster",
+				},
+			},
+			expectedConfig: Config{
+				ProjectID:   "project",
+				GCPLocation: "us-west1",
+				ClusterName: "cluster",
+			},
+		},
+		{
+			name:           "nil",
+			configMap:      nil,
+			expectedConfig: Config{},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			c, err := NewStackdriverConfigFromConfigMap(test.configMap)
+			if err != nil {
+				t.Errorf("Unexpected error: %v", err)
+			}
+			if test.expectedConfig != *c {
+				t.Errorf("Incorrect stackdriver config. Expected: [%v], Got: [%v]", test.expectedConfig, *c)
+			}
+		})
+	}
+}
+
+// This test ensures that ensureKubeClient completes. Errors are expected, but ok.
+func TestEnsureKubeclientNoDeadlock(t *testing.T) {
+	done := make(chan bool)
+	for i := 0; i < 10; i++ {
+		go testKubeclient(t, done)
+	}
+
+	// Wait for 10 responses
+	for i := 0; i < 10; i++ {
+		<-done
+	}
+}
```

Comment:
Pending

Reply:
Test has been simplified to just calling ensureKubeclient once and checking for panic, since otherwise it's just a test of sync.Once. This was a great tip for the future though.